src/randomization.ts

```diff
@@ -0,0 +1,350 @@
+/*
+ * (Psuedo-)Randomization functions for jsPsych
+ * Copyright (C) 2020  Ty Mees, Utrecht University
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+export {
+    randomizeStimuli,
+    randomizeStimuliConstraints,
+    randomShuffle,
+    randomShuffleConstraints,
+    stimuliMeetConstraints
+};
+
+const NOT_SHUFFLED_ERROR_MSG =
+    "Unable to shuffle according to the constraints, " +
+    "perhaps it is an idea to loosen the constraints.";
+
+
+/**
+ * Swaps two items in an array
+ * @param stimuli The array with objects/stimuli
+ * @param i1 a positive integer smaller than stimuli.length
+ * @param i2 a positive integer smaller than stimuli.length
+ */
+function swapItems<T>(stimuli: Array<T>, i1: number, i2: number) {
+    let temp = stimuli[i1];
+    stimuli[i1] = stimuli[i2];
+    stimuli[i2] = temp;
+}
+
+// A constraint a key (string) that is allowed a number of times in a row
+type ConstraintMapping = { [key:string]: number};
+
+type StringToUnknownMapping = {[key:string] : unknown};
```

Comment:
I'm not sure what you get when using this. Why not use `func<T>(stimuli: T[])` in all functions that use stimuli lists?

Reply:
Yes, that would be nice! However, when check whether we are going to check if stimulus.item_type has the same value for n times in a row, we "index" the items by key ({item_type: "TEST", ...} and Typescript can't verify that T is indexable by a string. So I'm trying to say using types that denote we have an array of:
```typescript
let stimluli = [
  { aap: 'bonobo', item_type: 'cond_a' },
  { aap: 'gorilla', item_type: 'cond_a' }
];
```
So when the constraints are ```{item_type: 2}```  The array above passes, because 2 * cond_a in a row is fine, but using ```{item_type: 1}``` as constraint would notify that the array above isn't meeting the constraints, hence we need to try again or return ```null``` as no valid order was found.

Reply:
You can still use a generic in this case, and I do think that would be better as that will preserve the item-typing in the output of the functions. 

Instead of typing a big response on how to that, I've implemented a proper generic type in this PR for illustration:
https://github.com/UiL-OTS-labs/ils-online-utils/pull/2